baseplate/url.py

```diff
@@ -0,0 +1,32 @@
+"""URL domain parsing."""
+
+
+import re
+
+DOMAIN_REGEX = re.compile("(?i)(?:.+?://)?([^/:#?]*)")
+DOMAIN_PREFIX_REGEX = re.compile(r"^www\d*\.")
+
+
+def is_subdomain(subdomain: str, base: str) -> bool:
+    """Check if a domain is equal to or a subdomain of a base domain."""
+    return subdomain == base or (subdomain is not None and subdomain.endswith("." + base))
+
+
+def get_domain(url: str) -> str:
+    """Take a URL and return the domain part (minus www.) if present."""
+    match = DOMAIN_REGEX.search(url)
```

Comment:
why do this with a regex instead of `urllib.parse`?

Reply:
This was initially copied from r2 into gateway and proliferated from there.  I _think_ using a regex is a little faster than using urllib?

```
In [52]: %timeit test(get_domain)                                                                           
3.13 µs ± 53.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [53]: %timeit test(urllib_domain)                                                                        
9.58 µs ± 71.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
```python
import urllib.parse

def urllib_domain(url): 
    return '.'.join(urllib.parse.urlparse(url).netloc.split('.')[-2:])

cases = []
tmpl = "https://%s.%s"
library = "abcdefghijklmnopqrstuvwxyz0123456789-"
endings = ['com', 'io', 'net', 'gov', 'test']

def build_case():
    ending = random.choice(endings)
    use_www = random.randint(0, 9) == 0
    parts = []
    if use_www:
        parts.append('www')
    domain = "".join(random.choice(library) for _ in range(random.randint(4, 15))) 
    subdomains = [
        "".join(random.choice(library) for _ in range(random.randint(4, 15)))
        for _ in range(random.randint(1, 3))
    ]
    parts.extend(subdomains)
    parts.append(domain)
    return tmpl % (".".join(parts), ending)

for _ in range(1000000):
    cases.append(build_case())

def test(fn):
    global cases
    case = random.choice(cases)
    fn(case)
```